pkg/reconciler/apis/apibinding/apibinding_controller.go

```diff
@@ -69,19 +69,50 @@ func NewController(
 	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
 
 	c := &controller{
-		queue:                    queue,
-		enqueueAfter:             func(binding *apisv1alpha1.APIBinding, duration time.Duration) { queue.AddAfter(binding, duration) },
-		crdClusterClient:         crdClusterClient,
-		kcpClusterClient:         kcpClusterClient,
-		apiBindingsLister:        apiBindingInformer.Lister(),
-		apiBindingsIndexer:       apiBindingInformer.Informer().GetIndexer(),
-		apiExportsLister:         apiExportInformer.Lister(),
-		apiExportsIndexer:        apiExportInformer.Informer().GetIndexer(),
-		apiResourceSchemaLister:  apiResourceSchemaInformer.Lister(),
+		queue:            queue,
+		enqueueAfter:     func(binding *apisv1alpha1.APIBinding, duration time.Duration) { queue.AddAfter(binding, duration) },
+		crdClusterClient: crdClusterClient,
+		kcpClusterClient: kcpClusterClient,
+
+		apiBindingsLister: apiBindingInformer.Lister(),
+		listAPIBindings: func(clusterName logicalcluster.LogicalCluster) ([]*apisv1alpha1.APIBinding, error) {
+			list, err := apiBindingInformer.Lister().List(labels.Everything())
+			if err != nil {
+				return nil, err
+			}
+
+			var ret []*apisv1alpha1.APIBinding
+
+			for i := range list {
+				if list[i].ClusterName != clusterName.String() {
+					continue
+				}
+
+				ret = append(ret, list[i])
+			}
+
+			return ret, nil
+		},
+		apiBindingsIndexer: apiBindingInformer.Informer().GetIndexer(),
+
+		getAPIExport: func(clusterName logicalcluster.LogicalCluster, name string) (*apisv1alpha1.APIExport, error) {
+			return apiExportInformer.Lister().Get(clusters.ToClusterAwareKey(clusterName, name))
+		},
+		apiExportsIndexer: apiExportInformer.Informer().GetIndexer(),
+
+		getAPIResourceSchema: func(clusterName logicalcluster.LogicalCluster, name string) (*apisv1alpha1.APIResourceSchema, error) {
+			return apiResourceSchemaInformer.Lister().Get(clusters.ToClusterAwareKey(clusterName, name))
+		},
 		apiResourceSchemaIndexer: apiResourceSchemaInformer.Informer().GetIndexer(),
-		crdLister:                crdInformer.Lister(),
-		crdIndexer:               crdInformer.Informer().GetIndexer(),
-		deletedCRDTracker:        newLockedStringSet(),
+
+		createCRD: func(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) (*apiextensionsv1.CustomResourceDefinition, error) {
+			return crdClusterClient.Cluster(logicalcluster.From(crd)).ApiextensionsV1().CustomResourceDefinitions().Create(ctx, crd, metav1.CreateOptions{})
```

Comment:
(no action necessary) am not a fan of implicitly passing the cluster name through `metadata.clusterName` here in code. IMO it is clear to have it explicitly everywhere.

Reply:
Updated to take in clusterName in incoming commit